extension/observer/hostobserver/extension.go

```diff
@@ -0,0 +1,253 @@
+// limitations under the License.
+
+package hostobserver
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"syscall"
+
+	"github.com/shirou/gopsutil/net"
+	"github.com/shirou/gopsutil/process"
+	"go.opentelemetry.io/collector/component"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+)
+
+type hostObserver struct {
+	observer.EndpointsWatcher
+}
+
+type endpointsLister struct {
+	logger       *zap.Logger
+	observerName string
+}
+
+var _ component.ServiceExtension = (*hostObserver)(nil)
+
+func newObserver(logger *zap.Logger, config *Config) (component.ServiceExtension, error) {
+	h := &hostObserver{
+		EndpointsWatcher: observer.EndpointsWatcher{
+			RefreshInterval: config.RefreshInterval,
+			Endpointslister: endpointsLister{
+				logger:       logger,
+				observerName: config.Name(),
+			},
+		},
+	}
+
+	return h, nil
+}
+
+func (h *hostObserver) Start(context.Context, component.Host) error {
+	return nil
+}
+
+func (h *hostObserver) Shutdown(context.Context) error {
+	h.StopListAndWatch()
+	return nil
+}
+
+func (e endpointsLister) ListEndpoints() []observer.Endpoint {
+	conns, err := getConnections()
+	if err != nil {
+		e.logger.Error("Could not get local network listeners", zap.Error(err))
+		return nil
+	}
+
+	return e.collectEndpoints(conns)
+}
+
+var getConnections = func() (conns []net.ConnectionStat, err error) {
+	// Skip UID lookup since it's not used by the observer, the method
+	// is available only on linux. See https://github.com/shirou/gopsutil/pull/783
+	// for details.
+	if runtime.GOOS == "linux" {
+		conns, err = net.ConnectionsWithoutUids("all")
+	} else {
+		conns, err = net.Connections("all")
+	}
+
+	return conns, err
+}
+
+func (e endpointsLister) collectEndpoints(conns []net.ConnectionStat) []observer.Endpoint {
+	endpoints := make([]observer.Endpoint, 0, len(conns))
+	connsByPID := make(map[int32][]*net.ConnectionStat)
+	for i := range conns {
+		c := conns[i]
+		isIPSocket := c.Family == syscall.AF_INET || c.Family == syscall.AF_INET6
+		isTCPOrUDP := c.Type == syscall.SOCK_STREAM || c.Type == syscall.SOCK_DGRAM
+		// UDP doesn't have any status
+		isUDPOrListening := c.Type == syscall.SOCK_DGRAM || c.Status == "LISTEN"
+		// UDP is "listening" when it has a remote port of 0
+		isTCPOrHasNoRemotePort := c.Type == syscall.SOCK_STREAM || c.Raddr.Port == 0
+
+		if !isIPSocket || !isTCPOrUDP || !isUDPOrListening || !isTCPOrHasNoRemotePort {
+			continue
+		}
+
+		// PID of 0 means that the listening file descriptor couldn't be mapped
+		// back to a process's set of open file descriptors in /proc. Collect these
+		// endpoints even though there's no process metadata available so users can
+		// still do discovery rules on such sockets.
+		if c.Pid == 0 {
+			cd := collectConnectionDetails(&c)
+			id := observer.EndpointID(
+				fmt.Sprintf(
+					"(%s)%s-%d-%s", e.observerName, c.Laddr.IP, cd.port, cd.transport,
```

Comment:
I'd like to make it and k8sobserver id format similar but there's some tweaks I want to make to id so I'll just do that separately in another PR.

I think it's better if we use metadata on the end of the ID, similar to what receiver_creator does:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/bb81db76bc6130e348ab6b2ab401d328b537f4ac/receiver/receivercreator/runner.go#L101

So this would look something like:

`hostobserver/2{endpoint="127.0.0.1:23", protocol="UDP", pid=3482}`

or something like that.

(Just as an aside RC should probably just use the endpoint ID to resolve that TODO)

tl;dr: nothing for you to change here right now, just notes on things to fix sometime.